python/observation-publisher/publisher/observation_sink.py

```diff
@@ -314,6 +320,236 @@ def write(self, dataframe: pd.DataFrame):
+            
+        except NoSuchObject:
+            create_table_query = "create table {table_name} ( {cols} )  auto partitioned by (trunc_time({timestamp_column}, 'day') as request_timestamp_pt)".format(
+                table_name=self.table_name_with_dataset,
+                cols=self.get_column_values_for_query(schema_fields=self.schema_fields),
+                timestamp_column=PREDICTION_LOG_TIMESTAMP_COLUMN
+            )
+            self._client.execute_sql(create_table_query)
+
+    @property
+    def schema_fields(self) -> List[Column]:
+        value_type_to_maxcompute_type = {
+            ValueType.INT64: "BIGINT",
+            ValueType.FLOAT64: "FLOAT",
+            ValueType.BOOLEAN: "BOOLEAN",
+            ValueType.STRING: "STRING",
+        }
+
+        schema_fields = [
+            Column(
+                name=self._inference_schema.session_id_column,
+                type="STRING",
+            ),
+            Column(
+                name=self._inference_schema.row_id_column,
+                type="STRING",
+            ),
+            Column(
+                name=PREDICTION_LOG_TIMESTAMP_COLUMN,
+                type="TIMESTAMP",
+            ),
+            Column(
+                name=PREDICTION_LOG_MODEL_VERSION_COLUMN,
+                type="STRING",
+            ),
+        ]
+        for feature, feature_type in self._inference_schema.feature_types.items():
+            schema_fields.append(
+                Column(
+                    name=feature, type=value_type_to_maxcompute_type[feature_type]
+                )
+            )
+        for (
+            prediction,
+            prediction_type,
+        ) in self._inference_schema.model_prediction_output.prediction_types().items():
+            schema_fields.append(
+                Column(
+                    name=prediction, type=value_type_to_maxcompute_type[prediction_type]
+                )
+            )
+
+        return schema_fields
+
+    @property
+    def write_location(self) -> str:
+        """
+        Returns the MaxCompute table location to write the prediction logs, which will be unique
+        for each CaraML project / model pair. Different versions of a model share the same table.
+        :return:
+        """
+        table_name = f"prediction_log_{self._project}_{self._model_id}".replace("-", "_").replace(
+            ".", "_"
+        )
+        return f"{self.max_compute_project}.{self.dataset}.{table_name}"
+    
+    @property
+    def table_name_with_dataset(self) -> str:
+        """
+        Returns the MaxCompute table location with the dataset prefix
+        for instance, the returned table name will be {dataset}.{table_name}
+        :return:
+        """
+        table_name = f"prediction_log_{self._project}_{self._model_id}".replace("-", "_").replace(
+            ".", "_"
+        )
+        return f"{self.dataset}.{table_name}"
+    
+    def get_schema_fields(self, schema) -> List[str]:
```

Comment:
We can add an underscore to the name of this method too, since I don't think users are expected to call this directly.